aiokafka/message_accumulator.py

```diff
@@ -0,0 +1,185 @@
+            # messages in async task
+            raise ProducerClosed()
+
+        batch = self._batches.get(tp)
+        if not batch:
+            message_set_buffer = MessageSetBuffer(
+                io.BytesIO(), self._batch_size, self._compression_type)
+            batch = MessageBatch(tp, message_set_buffer, self._ttl, self._loop)
+            self._batches[tp] = batch
+
+            efut = self._empty_futures.get(tp)
+            if efut is None or efut.done():
+                # batch per topic-partition is not empty
+                # so create new "empty" future
+                self._empty_futures[tp] = asyncio.Future(loop=self._loop)
+            if not self._wait_data_future.done():
+                # we have some data, so resolve "wait data" future
+                self._wait_data_future.set_result(None)
+
+        future = batch.append(key, value)
+        if future is None:
+            # waiting until batch per topic-partition is drained
+            done, _ = yield from asyncio.wait(
+                [self._empty_futures[tp]], timeout=self._ttl, loop=self._loop)
+            if not done:
+                raise KafkaTimeoutError()
+            return (yield from self.add_message(tp, key, value))
+        return future
+
+    @asyncio.coroutine
+    def wait_data(self):
+        """wait until there are some data for send"""
+        if self._wait_data_future.done():
+            return
+        yield from self._wait_data_future
+
+    @asyncio.coroutine
+    def flush(self):
+        """wait until all batches is drained by send task"""
+        for batch in list(self._batches.values()):
+            yield from batch.wait_all()
+
+    def _pop_batch(self, tp):
+        batch = self._batches.pop(tp)
+        batch.pack()
+        if not self._empty_futures[tp].done():
+            self._empty_futures[tp].set_result(None)
+        return batch
+
+    def drain_by_nodes(self, ignore_nodes):
+        """return batches by nodes"""
+        nodes = collections.defaultdict(dict)
+        unknown_leaders_exist = False
+        for tp in list(self._batches.keys()):
+            leader = self._cluster.leader_for_partition(tp)
+            if leader is None or leader == -1:
+                if self._batches[tp].expired():
+                    # batch is for partition is expired and still no leader,
+                    # so set exception for batch and pop it
+                    batch = self._pop_batch(tp)
+                    if leader is None:
+                        err = NotLeaderForPartitionError()
+                    else:
+                        err = LeaderNotAvailableError()
+                    batch.done(exception=err)
+                unknown_leaders_exist = True
+                continue
+            elif ignore_nodes and leader in ignore_nodes:
+                continue
+
+            batch = self._pop_batch(tp)
+            nodes[leader][tp] = batch
+
+        if not self._batches:
```

Comment:
I think this would be better to do in `wait_data` method. We will only call it 1-ce anyway